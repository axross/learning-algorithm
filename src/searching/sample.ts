export const sortedCharactors = [
  "a",
  "b",
  "c",
  "d",
  "e",
  "f",
  "g",
  "h",
  "i",
  "j",
  "k",
  "l",
  "m",
  "n",
  "o",
  "p",
  "q",
  "r",
  "s",
  "t",
  "u",
  "v",
  "w",
  "x",
  "y",
  "z"
];

export const unsortedCharactors = [
  "c",
  "n",
  "r",
  "j",
  "m",
  "s",
  "b",
  "u",
  "w",
  "y",
  "h",
  "t",
  "v",
  "f",
  "z",
  "g",
  "x",
  "o",
  "k",
  "q",
  "d",
  "e",
  "p",
  "l",
  "a",
  "i"
];
