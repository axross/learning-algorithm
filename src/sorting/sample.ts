import roundRobin from "../test-utility/round-robin.ts";
import randomInteger from "../test-utility/random-integer.ts";

export const charactors = ["a", "b", "c", "d", "e", "f", "g", "h"];

export const samples = roundRobin(charactors).slice(1);

export function getRandomSample() {
  return samples[randomInteger(0, samples.length - 1)];
}

export const staticSample = ["f", "a", "h", "b", "d", "g", "e", "c"];

export const charactorPairs = ["a0", "a1", "b0", "b1", "c0", "c1", "d0", "d1"];

export const charactorPairSamples = roundRobin(["a0", "b0", "c0", "d0"]).reduce(
  (arr, pattern0) => [
    ...arr,
    ...roundRobin(["a1", "b1", "c1", "d1"]).reduce(
      (arr, pattern1) => [...arr, [...pattern0, ...pattern1]],
      [] as string[][]
    )
  ],
  [] as string[][]
);

export const randomCharactorPairSample =
  charactorPairSamples[randomInteger(0, charactorPairSamples.length - 1)];

export const staticCharactorPairSample =
  randomCharactorPairSample[Math.floor((samples.length / 7) * 5)];
