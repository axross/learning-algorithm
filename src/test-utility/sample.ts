function shuffle<Value>(arr: Value[]): Value[] {
  const shuffled = [...arr];

  for (let i = shuffled.length - 1; i >= 0; i--) {
    const rand = Math.floor(Math.random() * (i + 1));

    [shuffled[i], shuffled[rand]] = [shuffled[rand], shuffled[i]];
  }

  return shuffled;
}

export const sampleCharactors = [
  ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 65)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 97))
];

export const shuffledSampleCharactors = shuffle(sampleCharactors);

export const sampleTwoCharactors = [
  ...Array.from({ length: 5 }, (_, i) => [
    String.fromCharCode(i + 48) + "<",
    String.fromCharCode(i + 48) + ">"
  ]),
  ...Array.from({ length: 13 }, (_, i) => [
    String.fromCharCode(i + 65) + "<",
    String.fromCharCode(i + 65) + ">"
  ]),
  ...Array.from({ length: 13 }, (_, i) => [
    String.fromCharCode(i + 97) + "<",
    String.fromCharCode(i + 97) + ">"
  ])
].reduce((array, value) => [...array, ...value], []);

export const shuffledSampleTwoCharactors = [
  ...shuffle([
    ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48) + "<"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 65) + "<"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 97) + "<")
  ]),
  ...shuffle([
    ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48) + ">"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 65) + ">"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 97) + ">")
  ])
];
