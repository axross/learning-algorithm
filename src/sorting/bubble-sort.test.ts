import bubbleSort from "./bubble-sort.ts";
import { charactors, getRandomSample, staticSample } from "./sample.ts";
import { test, assert } from "https://deno.land/x/testing/mod.ts";

test({
  name: "bubbleSort() sorts any array to be ordered",
  fn() {
    for (let i = 1; i <= 100; ++i) {
      const array = [...getRandomSample()];

      bubbleSort({
        array,
        compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0)
      });

      assert.equal(array, charactors);
    }
  }
});

test({
  name: "bubbleSort() sorts an array in the correct process",
  fn() {
    const changesOnSwap: string[][] = [];
    const array = [...staticSample];

    bubbleSort({
      array,
      compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0),
      onSwap: () => changesOnSwap.push([...array])
    });

    assert.equal(changesOnSwap, [
      ["a", "f", "h", "b", "d", "g", "e", "c"],
      ["a", "f", "b", "h", "d", "g", "e", "c"],
      ["a", "f", "b", "d", "h", "g", "e", "c"],
      ["a", "f", "b", "d", "g", "h", "e", "c"],
      ["a", "f", "b", "d", "g", "e", "h", "c"],
      ["a", "f", "b", "d", "g", "e", "c", "h"],
      ["a", "b", "f", "d", "g", "e", "c", "h"],
      ["a", "b", "d", "f", "g", "e", "c", "h"],
      ["a", "b", "d", "f", "e", "g", "c", "h"],
      ["a", "b", "d", "f", "e", "c", "g", "h"],
      ["a", "b", "d", "e", "f", "c", "g", "h"],
      ["a", "b", "d", "e", "c", "f", "g", "h"],
      ["a", "b", "d", "c", "e", "f", "g", "h"],
      ["a", "b", "c", "d", "e", "f", "g", "h"]
    ]);
  }
});
