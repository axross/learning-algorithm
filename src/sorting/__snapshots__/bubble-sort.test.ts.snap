// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the steps of bubbleSort() matches the previous snapshot 1`] = `
Array [
  Object {
    "a": Object {
      "index": 1,
      "value": "z",
    },
    "b": Object {
      "index": 2,
      "value": "l",
    },
  },
  Object {
    "a": Object {
      "index": 2,
      "value": "z",
    },
    "b": Object {
      "index": 3,
      "value": "j",
    },
  },
  Object {
    "a": Object {
      "index": 3,
      "value": "z",
    },
    "b": Object {
      "index": 4,
      "value": "k",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "z",
    },
    "b": Object {
      "index": 5,
      "value": "y",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "z",
    },
    "b": Object {
      "index": 6,
      "value": "g",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "z",
    },
    "b": Object {
      "index": 7,
      "value": "b",
    },
  },
  Object {
    "a": Object {
      "index": 7,
      "value": "z",
    },
    "b": Object {
      "index": 8,
      "value": "s",
    },
  },
  Object {
    "a": Object {
      "index": 8,
      "value": "z",
    },
    "b": Object {
      "index": 9,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 9,
      "value": "z",
    },
    "b": Object {
      "index": 10,
      "value": "x",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "z",
    },
    "b": Object {
      "index": 11,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "z",
    },
    "b": Object {
      "index": 12,
      "value": "i",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "z",
    },
    "b": Object {
      "index": 13,
      "value": "v",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "z",
    },
    "b": Object {
      "index": 14,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 14,
      "value": "z",
    },
    "b": Object {
      "index": 15,
      "value": "n",
    },
  },
  Object {
    "a": Object {
      "index": 15,
      "value": "z",
    },
    "b": Object {
      "index": 16,
      "value": "q",
    },
  },
  Object {
    "a": Object {
      "index": 16,
      "value": "z",
    },
    "b": Object {
      "index": 17,
      "value": "t",
    },
  },
  Object {
    "a": Object {
      "index": 17,
      "value": "z",
    },
    "b": Object {
      "index": 18,
      "value": "p",
    },
  },
  Object {
    "a": Object {
      "index": 18,
      "value": "z",
    },
    "b": Object {
      "index": 19,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 19,
      "value": "z",
    },
    "b": Object {
      "index": 20,
      "value": "u",
    },
  },
  Object {
    "a": Object {
      "index": 20,
      "value": "z",
    },
    "b": Object {
      "index": 21,
      "value": "w",
    },
  },
  Object {
    "a": Object {
      "index": 21,
      "value": "z",
    },
    "b": Object {
      "index": 22,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 22,
      "value": "z",
    },
    "b": Object {
      "index": 23,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 23,
      "value": "z",
    },
    "b": Object {
      "index": 24,
      "value": "r",
    },
  },
  Object {
    "a": Object {
      "index": 24,
      "value": "z",
    },
    "b": Object {
      "index": 25,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 0,
      "value": "o",
    },
    "b": Object {
      "index": 1,
      "value": "l",
    },
  },
  Object {
    "a": Object {
      "index": 1,
      "value": "o",
    },
    "b": Object {
      "index": 2,
      "value": "j",
    },
  },
  Object {
    "a": Object {
      "index": 2,
      "value": "o",
    },
    "b": Object {
      "index": 3,
      "value": "k",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "y",
    },
    "b": Object {
      "index": 5,
      "value": "g",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "y",
    },
    "b": Object {
      "index": 6,
      "value": "b",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "y",
    },
    "b": Object {
      "index": 7,
      "value": "s",
    },
  },
  Object {
    "a": Object {
      "index": 7,
      "value": "y",
    },
    "b": Object {
      "index": 8,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 8,
      "value": "y",
    },
    "b": Object {
      "index": 9,
      "value": "x",
    },
  },
  Object {
    "a": Object {
      "index": 9,
      "value": "y",
    },
    "b": Object {
      "index": 10,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "y",
    },
    "b": Object {
      "index": 11,
      "value": "i",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "y",
    },
    "b": Object {
      "index": 12,
      "value": "v",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "y",
    },
    "b": Object {
      "index": 13,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "y",
    },
    "b": Object {
      "index": 14,
      "value": "n",
    },
  },
  Object {
    "a": Object {
      "index": 14,
      "value": "y",
    },
    "b": Object {
      "index": 15,
      "value": "q",
    },
  },
  Object {
    "a": Object {
      "index": 15,
      "value": "y",
    },
    "b": Object {
      "index": 16,
      "value": "t",
    },
  },
  Object {
    "a": Object {
      "index": 16,
      "value": "y",
    },
    "b": Object {
      "index": 17,
      "value": "p",
    },
  },
  Object {
    "a": Object {
      "index": 17,
      "value": "y",
    },
    "b": Object {
      "index": 18,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 18,
      "value": "y",
    },
    "b": Object {
      "index": 19,
      "value": "u",
    },
  },
  Object {
    "a": Object {
      "index": 19,
      "value": "y",
    },
    "b": Object {
      "index": 20,
      "value": "w",
    },
  },
  Object {
    "a": Object {
      "index": 20,
      "value": "y",
    },
    "b": Object {
      "index": 21,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 21,
      "value": "y",
    },
    "b": Object {
      "index": 22,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 22,
      "value": "y",
    },
    "b": Object {
      "index": 23,
      "value": "r",
    },
  },
  Object {
    "a": Object {
      "index": 23,
      "value": "y",
    },
    "b": Object {
      "index": 24,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 0,
      "value": "l",
    },
    "b": Object {
      "index": 1,
      "value": "j",
    },
  },
  Object {
    "a": Object {
      "index": 1,
      "value": "l",
    },
    "b": Object {
      "index": 2,
      "value": "k",
    },
  },
  Object {
    "a": Object {
      "index": 3,
      "value": "o",
    },
    "b": Object {
      "index": 4,
      "value": "g",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "o",
    },
    "b": Object {
      "index": 5,
      "value": "b",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "s",
    },
    "b": Object {
      "index": 7,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 8,
      "value": "x",
    },
    "b": Object {
      "index": 9,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 9,
      "value": "x",
    },
    "b": Object {
      "index": 10,
      "value": "i",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "x",
    },
    "b": Object {
      "index": 11,
      "value": "v",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "x",
    },
    "b": Object {
      "index": 12,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "x",
    },
    "b": Object {
      "index": 13,
      "value": "n",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "x",
    },
    "b": Object {
      "index": 14,
      "value": "q",
    },
  },
  Object {
    "a": Object {
      "index": 14,
      "value": "x",
    },
    "b": Object {
      "index": 15,
      "value": "t",
    },
  },
  Object {
    "a": Object {
      "index": 15,
      "value": "x",
    },
    "b": Object {
      "index": 16,
      "value": "p",
    },
  },
  Object {
    "a": Object {
      "index": 16,
      "value": "x",
    },
    "b": Object {
      "index": 17,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 17,
      "value": "x",
    },
    "b": Object {
      "index": 18,
      "value": "u",
    },
  },
  Object {
    "a": Object {
      "index": 18,
      "value": "x",
    },
    "b": Object {
      "index": 19,
      "value": "w",
    },
  },
  Object {
    "a": Object {
      "index": 19,
      "value": "x",
    },
    "b": Object {
      "index": 20,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 20,
      "value": "x",
    },
    "b": Object {
      "index": 21,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 21,
      "value": "x",
    },
    "b": Object {
      "index": 22,
      "value": "r",
    },
  },
  Object {
    "a": Object {
      "index": 22,
      "value": "x",
    },
    "b": Object {
      "index": 23,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 2,
      "value": "l",
    },
    "b": Object {
      "index": 3,
      "value": "g",
    },
  },
  Object {
    "a": Object {
      "index": 3,
      "value": "l",
    },
    "b": Object {
      "index": 4,
      "value": "b",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "o",
    },
    "b": Object {
      "index": 6,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 7,
      "value": "s",
    },
    "b": Object {
      "index": 8,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 8,
      "value": "s",
    },
    "b": Object {
      "index": 9,
      "value": "i",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "v",
    },
    "b": Object {
      "index": 11,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "v",
    },
    "b": Object {
      "index": 12,
      "value": "n",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "v",
    },
    "b": Object {
      "index": 13,
      "value": "q",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "v",
    },
    "b": Object {
      "index": 14,
      "value": "t",
    },
  },
  Object {
    "a": Object {
      "index": 14,
      "value": "v",
    },
    "b": Object {
      "index": 15,
      "value": "p",
    },
  },
  Object {
    "a": Object {
      "index": 15,
      "value": "v",
    },
    "b": Object {
      "index": 16,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 16,
      "value": "v",
    },
    "b": Object {
      "index": 17,
      "value": "u",
    },
  },
  Object {
    "a": Object {
      "index": 18,
      "value": "w",
    },
    "b": Object {
      "index": 19,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 19,
      "value": "w",
    },
    "b": Object {
      "index": 20,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 20,
      "value": "w",
    },
    "b": Object {
      "index": 21,
      "value": "r",
    },
  },
  Object {
    "a": Object {
      "index": 21,
      "value": "w",
    },
    "b": Object {
      "index": 22,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 1,
      "value": "k",
    },
    "b": Object {
      "index": 2,
      "value": "g",
    },
  },
  Object {
    "a": Object {
      "index": 2,
      "value": "k",
    },
    "b": Object {
      "index": 3,
      "value": "b",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "l",
    },
    "b": Object {
      "index": 5,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "o",
    },
    "b": Object {
      "index": 7,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 7,
      "value": "o",
    },
    "b": Object {
      "index": 8,
      "value": "i",
    },
  },
  Object {
    "a": Object {
      "index": 9,
      "value": "s",
    },
    "b": Object {
      "index": 10,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "s",
    },
    "b": Object {
      "index": 11,
      "value": "n",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "s",
    },
    "b": Object {
      "index": 12,
      "value": "q",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "t",
    },
    "b": Object {
      "index": 14,
      "value": "p",
    },
  },
  Object {
    "a": Object {
      "index": 14,
      "value": "t",
    },
    "b": Object {
      "index": 15,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 17,
      "value": "v",
    },
    "b": Object {
      "index": 18,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 18,
      "value": "v",
    },
    "b": Object {
      "index": 19,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 19,
      "value": "v",
    },
    "b": Object {
      "index": 20,
      "value": "r",
    },
  },
  Object {
    "a": Object {
      "index": 20,
      "value": "v",
    },
    "b": Object {
      "index": 21,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 0,
      "value": "j",
    },
    "b": Object {
      "index": 1,
      "value": "g",
    },
  },
  Object {
    "a": Object {
      "index": 1,
      "value": "j",
    },
    "b": Object {
      "index": 2,
      "value": "b",
    },
  },
  Object {
    "a": Object {
      "index": 3,
      "value": "k",
    },
    "b": Object {
      "index": 4,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "l",
    },
    "b": Object {
      "index": 6,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "l",
    },
    "b": Object {
      "index": 7,
      "value": "i",
    },
  },
  Object {
    "a": Object {
      "index": 8,
      "value": "o",
    },
    "b": Object {
      "index": 9,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 9,
      "value": "o",
    },
    "b": Object {
      "index": 10,
      "value": "n",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "s",
    },
    "b": Object {
      "index": 13,
      "value": "p",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "s",
    },
    "b": Object {
      "index": 14,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 16,
      "value": "u",
    },
    "b": Object {
      "index": 17,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 17,
      "value": "u",
    },
    "b": Object {
      "index": 18,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 18,
      "value": "u",
    },
    "b": Object {
      "index": 19,
      "value": "r",
    },
  },
  Object {
    "a": Object {
      "index": 19,
      "value": "u",
    },
    "b": Object {
      "index": 20,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 0,
      "value": "g",
    },
    "b": Object {
      "index": 1,
      "value": "b",
    },
  },
  Object {
    "a": Object {
      "index": 2,
      "value": "j",
    },
    "b": Object {
      "index": 3,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "k",
    },
    "b": Object {
      "index": 5,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "k",
    },
    "b": Object {
      "index": 6,
      "value": "i",
    },
  },
  Object {
    "a": Object {
      "index": 7,
      "value": "l",
    },
    "b": Object {
      "index": 8,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "q",
    },
    "b": Object {
      "index": 12,
      "value": "p",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "q",
    },
    "b": Object {
      "index": 13,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 15,
      "value": "t",
    },
    "b": Object {
      "index": 16,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 16,
      "value": "t",
    },
    "b": Object {
      "index": 17,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 17,
      "value": "t",
    },
    "b": Object {
      "index": 18,
      "value": "r",
    },
  },
  Object {
    "a": Object {
      "index": 18,
      "value": "t",
    },
    "b": Object {
      "index": 19,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 1,
      "value": "g",
    },
    "b": Object {
      "index": 2,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 3,
      "value": "j",
    },
    "b": Object {
      "index": 4,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "j",
    },
    "b": Object {
      "index": 5,
      "value": "i",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "k",
    },
    "b": Object {
      "index": 7,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "p",
    },
    "b": Object {
      "index": 12,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 14,
      "value": "s",
    },
    "b": Object {
      "index": 15,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 15,
      "value": "s",
    },
    "b": Object {
      "index": 16,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 16,
      "value": "s",
    },
    "b": Object {
      "index": 17,
      "value": "r",
    },
  },
  Object {
    "a": Object {
      "index": 17,
      "value": "s",
    },
    "b": Object {
      "index": 18,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 0,
      "value": "b",
    },
    "b": Object {
      "index": 1,
      "value": "a",
    },
  },
  Object {
    "a": Object {
      "index": 2,
      "value": "g",
    },
    "b": Object {
      "index": 3,
      "value": "d",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "j",
    },
    "b": Object {
      "index": 6,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "o",
    },
    "b": Object {
      "index": 11,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "q",
    },
    "b": Object {
      "index": 14,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 14,
      "value": "q",
    },
    "b": Object {
      "index": 15,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 16,
      "value": "r",
    },
    "b": Object {
      "index": 17,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "i",
    },
    "b": Object {
      "index": 5,
      "value": "h",
    },
  },
  Object {
    "a": Object {
      "index": 9,
      "value": "n",
    },
    "b": Object {
      "index": 10,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "p",
    },
    "b": Object {
      "index": 13,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "p",
    },
    "b": Object {
      "index": 14,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 15,
      "value": "q",
    },
    "b": Object {
      "index": 16,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 8,
      "value": "l",
    },
    "b": Object {
      "index": 9,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "o",
    },
    "b": Object {
      "index": 12,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "o",
    },
    "b": Object {
      "index": 13,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 14,
      "value": "p",
    },
    "b": Object {
      "index": 15,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 7,
      "value": "k",
    },
    "b": Object {
      "index": 8,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "n",
    },
    "b": Object {
      "index": 11,
      "value": "m",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "n",
    },
    "b": Object {
      "index": 12,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 13,
      "value": "o",
    },
    "b": Object {
      "index": 14,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "j",
    },
    "b": Object {
      "index": 7,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "m",
    },
    "b": Object {
      "index": 11,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 12,
      "value": "n",
    },
    "b": Object {
      "index": 13,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "i",
    },
    "b": Object {
      "index": 6,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 9,
      "value": "l",
    },
    "b": Object {
      "index": 10,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 11,
      "value": "m",
    },
    "b": Object {
      "index": 12,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "h",
    },
    "b": Object {
      "index": 5,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 8,
      "value": "k",
    },
    "b": Object {
      "index": 9,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 10,
      "value": "l",
    },
    "b": Object {
      "index": 11,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 3,
      "value": "g",
    },
    "b": Object {
      "index": 4,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 7,
      "value": "j",
    },
    "b": Object {
      "index": 8,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 9,
      "value": "k",
    },
    "b": Object {
      "index": 10,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 2,
      "value": "d",
    },
    "b": Object {
      "index": 3,
      "value": "c",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "i",
    },
    "b": Object {
      "index": 7,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 8,
      "value": "j",
    },
    "b": Object {
      "index": 9,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "h",
    },
    "b": Object {
      "index": 6,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 7,
      "value": "i",
    },
    "b": Object {
      "index": 8,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 4,
      "value": "g",
    },
    "b": Object {
      "index": 5,
      "value": "e",
    },
  },
  Object {
    "a": Object {
      "index": 6,
      "value": "h",
    },
    "b": Object {
      "index": 7,
      "value": "f",
    },
  },
  Object {
    "a": Object {
      "index": 5,
      "value": "g",
    },
    "b": Object {
      "index": 6,
      "value": "f",
    },
  },
]
`;
